chrome/content/zotero/elements/itemBox.js

```diff
@@ -866,13 +882,255 @@
+			if (!this._customRowElemCache) return;
+			for (let position of Object.keys(this._customRowElemCache)) {
+				this._customRowElemCache[position].forEach((rowElem) => {
+					this.insertCustomRow(rowElem, position);
+				});
+				this._customRowElemCache[position] = [];
+			}
+		}
+
+		insertCustomRow(rowElem, position = "end") {
+			switch (position) {
+				case "start": {
+					this._infoTable.prepend(rowElem);
+					break;
+				}
+				case "afterCreators": {
+					// The `_firstRowBeforeCreators` is actually the first row after creator rows
+					if (this._firstRowBeforeCreators) {
+						this._infoTable.insertBefore(rowElem, this._firstRowBeforeCreators);
+						// Update the anchor node for creator rows
+						this._firstRowBeforeCreators = rowElem;
+					}
+					else {
+						this._infoTable.append(rowElem);
+					}
+					break;
+				}
+				case "end":
+				default: {
+					let dateAddedRow = this._infoTable.querySelector(".meta-label[fieldname=dateAdded]")?.parentElement;
+					if (dateAddedRow) {
+						this._infoTable.insertBefore(rowElem, dateAddedRow);
+					}
+					else {
+						this._infoTable.append(rowElem);
+					}
+					break;
+				}
+			}
+		}
+		
+		updateCustomRowProperty(rowElem) {
+			if (!this.item) return;
+			let rowID = rowElem.dataset.customRowId;
+			if (!rowID) return;
+			let valueElem = rowElem.querySelector(".meta-data > .value");
+			if (!this.editable) valueElem.toggleAttribute('readonly', true);
+			
+			let onItemChange = Zotero.ItemPaneManager.getInfoRowHook(rowID, "onItemChange");
+			if (!onItemChange) return;
+			try {
+				onItemChange({
+					rowID,
+					item: this.item,
+					tabType: this.tabType,
+					editable: this.editable,
+					setEnabled: (enabled) => {
+						rowElem.hidden = !enabled;
+					},
+					setEditable: (editable) => {
+						if (!this.editable) editable = false;
+						rowElem.querySelector(".meta-data > .value").toggleAttribute('readonly', !editable);
+					}
+				});
+			}
+			catch (e) {
+				Zotero.logError(e);
+			}
+		}
+
+		updateCustomRowData(rowElem) {
+			if (!this.item) return;
+			let rowID = rowElem.dataset.customRowId;
+			let onGetData = Zotero.ItemPaneManager.getInfoRowHook(rowID, "onGetData");
+			if (!onGetData) return;
+			let data = "";
+			try {
+				data = onGetData({
+					rowID,
+					item: this.item,
+					tabType: this.tabType,
+					editable: this.editable,
+				});
+				if (typeof data !== "string") {
+					throw new Error("ItemPaneInfoRow onGetData must return a string");
+				}
+			}
+			catch (e) {
+				Zotero.logError(e);
+			}
+			let valueElem = rowElem.querySelector(".meta-data > .value");
+			valueElem.value = data;
+
+			// Attempt to make bidi things work automatically:
+			// If we have text to work off of, let the layout engine try to guess the text direction
+			if (data) {
+				valueElem.dir = 'auto';
+			}
+			// If not, assume it follows the locale's direction
+			else {
+				valueElem.dir = Zotero.dir;
+			}
```

Comment:
More duplicated code. I'd really like to try to avoid this if possible. It seems like all rows should be rendered in only one place, and if it's a custom row the existing functions should just get the necessary attributes and values.

Reply:
Agree. Back to the time I implemented this PR, I initially made some big refactors to the itemBox.js to unify all row-rendering calls, which have been a bit messy for a long time. However, these changes frequently conflict with commits in the main branch and make implementing the API based on those changes impossible. So I decided to duplicate the code a bit temporarily. Since the info box is no longer having big changes now, I think it's time to do this refactor again!